Properly support multiple keys for the same type of commodity

ading1977 opened the pull request:
https://vmturbo.atlassian.net/browse/OM-64661

Currently in `DIFEntity`, the `Metrics` is a map where the key is the metric type, and the value is an array of metric definitions:

```
type DIFEntity struct {
	UID                 string                     `json:"uniqueId"`
	Type                string                     `json:"type"`
	Name                string                     `json:"name"`
	HostedOn            *DIFHostedOn               `json:"hostedOn"`
	MatchingIdentifiers *DIFMatchingIdentifiers    `json:"matchIdentifiers"`
	PartOf              []*DIFPartOf               `json:"partOf"`
	Metrics             map[string][]*DIFMetricVal `json:"metrics"`
	namespace           string
	partOfSet           set.Set
	hostTypeSet         set.Set
}
```
We did not properly specify the use case for the array of metric definitions. After discussion with @pallavidn, we clarified that the only need for an array of metrics of the same type would be for those metrics with different keys.

This fix modifies the `AddMetric` SDK function to allow adding metrics of the same type but different keys for the same entity.

As an example, the metric created from using this SDK function may look like this when marshalled into JSON:

```JSON
kpi: [
    {
        average: 123,
        capacity: 1000,
        key: "total_messages_in_queue"
    },
    {
        average: 104.44444444444444,
        capacity: 1000,
        key: "total_waiting_time_in_queue"
    }
],
```

chlam4 commented on pkg/dataingestionframework/data/dif_types.go:
```diff
@@ -116,21 +116,40 @@ func (e *DIFEntity) Matching(id string) *DIFEntity {
-		return
+		// This is a new metric type
+		metricVal = &DIFMetricVal{Key: metricKey}
+		e.Metrics[metricType] = append(e.Metrics[metricType], metricVal)
```
Nit: Is `e.Metrics[metricType]` an empty slice in this case?  So probably doesn't need to use append at all?

chlam4 commented on pkg/dataingestionframework/data/dif_types.go:
```diff
@@ -116,21 +116,40 @@ func (e *DIFEntity) Matching(id string) *DIFEntity {
+			e.Metrics[metricType] = append(e.Metrics[metricType], metricVal)
+		}
+	} else {
+		if len(meList) < 1 {
```
Since this is a return, we can probably write it like the following which sounds clearer?
```
    } else if len(meList) < 1 {
       return
    } else {
       metricVal = meList[0]
    }
```

chlam4 replied:
Also suggest to add a comment for each of these if-else cases like you did at line 131 to explain the case.  It doesn't seem obvious to me at the first read, and I have to go thru a few times before understanding it better - but of course that's probably because I'm not that familiar with the code :)

chlam4 commented on pkg/dataingestionframework/data/dif_types.go:
```diff
@@ -116,21 +116,40 @@ func (e *DIFEntity) Matching(id string) *DIFEntity {
+		}
+	} else {
+		if len(meList) < 1 {
+			return
```
Actually should we return?  This is the case that it is found, but the value is an empty list.  I think that shouldn't happen, but if it does, shouldn't we add that new element to the list?

chlam4 commented on pkg/dataingestionframework/data/dif_types_test.go:
```diff
@@ -0,0 +1,56 @@
+	assert.Nil(t, metric.Key)
+}
+
+func TestAddMetricWithSameKey(t *testing.T) {
```
can we add some tests for overwriting the value and ensure that works too?

ading1977 commented:
Addressed all the review comments. Refactored the code to:

* Add more detailed comments
* Refactor the logic to make it cleaner
* Add unit test to test overwriting existing metric values
